### ISSUE TITLE
Devs: STATIC_FILTER code problems

### ISSUE THREAD

**Issue by JayDi85** [magefree/mage#4402]
Static filters is not static and can be changed anytime by source code. Well, it's not only static filters problem, but other "static" code affected too ([early example with mana choices](https://github.com/magefree/mage/commit/75cec526dcd47fe26e42b79c63179a8dd505c5c6)).

Last example with broken tests for Emrakul, the Aeons Torn (issue #4389, fix #4401). It's work fine until some card is not use dirty ability and change that static filter (add more predicates to filter):

Broken static filter with dirty data after batch of tests:
![shot_180112_151110](https://user-images.githubusercontent.com/8344157/34872635-1d7d985e-f7ab-11e7-9ab2-dd21fba5d6f2.png)

How it can be modified - example:
![shot_180112_152937](https://user-images.githubusercontent.com/8344157/34873164-7aac370e-f7ad-11e7-8cc8-cc03b15bd6fd.png)

How to solve that problem:
 1. Need to use filter.copy() all around code;
 2. Somehow "lock" filter changed and throw exception on add new data to it:
```
public static final FilterControlledPermanent FILTER_XXX = new FilterControlledPermanent();
static {
  FILTER_XXX.add(Predicates.not(new CardTypePredicate(CardType.LAND)));
  FILTER_XXX.lockForChanges(true);
}
```
```
    @Override
    public final Filter add(Predicate predicate) {

        if (this.changesLocked) {
            throw new IllegalStateException("Filter was locked for any changes");
        }

        predicates.add(predicate);
        return this;
    }
```

Result:
![shot_180112_155913](https://user-images.githubusercontent.com/8344157/34874230-98e7f678-f7b1-11e7-919d-92c865546132.png)

**Comment by rsypen** [comment]
Issues with your approaches:
1. Need to use filter.copy() all around code;
**Yes, but it requires changes everywhere + there is possiblity to copy already modified filter...**
2. Somehow "lock" filter changed and throw exception on add new data to it
**It just overcomplicates implementation of Filters and it's bad design in overall.**

I think better options are:
1. Replace static fields with static methods producing new filter objects instead of returning shared instances.
2. Prohibit modifications to Filter classes:
- Remove methods that allow to change state of Filter or
- Change signatures of methods so that changing state always produce new fresh instance of Filter (chaining) with change.

**Comment by rsypen** [comment]
Personally I'd go with **Prohibit modifications to Filter classes** because concept of Filter function is that it should not really allow state modifications because making change to Filter gives you completely new filter. That concept should be reflected in code.

**Comment by JayDi85** [comment]
> Prohibit modifications to Filter classes

You can't do it. Literally every cards use filter modification:
```
    private static final FilterCreaturePermanent filter = new FilterCreaturePermanent("Ally creatures you control");
    static {
        filter.add(new SubtypePredicate(SubType.ALLY));
        filter.add(new ControllerPredicate(TargetController.YOU));
    }
```
> 1.Need to use filter.copy() all around code;
> Yes, but it requires changes everywhere + there is possiblity to copy already modified filter...

Only some cards have dirty code and modify static filters.

**Comment by ingmargoudt** [comment]
I can understand that a static filter is 'extended' when a card class also has a static initializer that falsely makes use of that filter. Which cards do have this 'dirty code' ?

**Comment by JayDi85** [comment]
Pucas Mischief

**Comment by JayDi85** [comment]
Also some filters construct dynamical. Example with Raging River:

![shot_180112_180008](https://user-images.githubusercontent.com/8344157/34878130-7a10320e-f7c2-11e7-8ca1-af471278b323.png)

**Comment by ingmargoudt** [comment]
Ahhh s***t !!! 
thats my bad!! I changed a code in TargetControlledPermanent.java wrong to use a static filter!! And this ripples down to TargetPermanent that has the filter as class attribute. So that explains why the filter is changed at runtime

**Comment by ingmargoudt** [comment]
TargetControlledPermanent.java::42 and ::46. This should be a 'new FilterControlledPermanent()'  instead of the Static Filter reference.

**Comment by ingmargoudt** [comment]
The filter used in Raging river is fine, as that is local to the method and cleaned up afterwards by gc.

**Comment by ingmargoudt** [comment]
I will revert the Static Filter reference in TargetControlledPermanent

**Comment by JayDi85** [comment]
@ingmargoudt no, it's not your fault. It's a static filter at all. Don't change it.

**Comment by ingmargoudt** [comment]
Speaking of Raging River... I see object comparison using == rather than .equals()...

**Comment by rsypen** [comment]
We can make filter class immutable and add method like:

createModificableCopy(): MutableFilter

and provide all methods that changes state to MutableFilter class. MutableFilter class will be wrapper for any Filter instance that will operate on copy of filter.

Example.
MutableFilter mutable = StaticFilter.SOME_FILTER.createModificableCopy();
mutable.add(Predicate)
if(...) {
  if(..) {
    mutable.add(Predicate)
  }
}
battlefield.count(mutable)

**Comment by ingmargoudt** [comment]
The StaticFilters were introduced to reduce code duplication, so instead of new ..Filter() references we just used the StaticFilter instance. I just don't see where in the code of Pucas Mischief a Static Filter is changed.

@JayDi85 I do see now that the staticfilter I used was indeed correct, since it was never changed anyway...

**Comment by ingmargoudt** [comment]
Same for 'TargetPermanent'

**Comment by rsypen** [comment]
@ingmargoudt, I understand why they were introduced but problem is that you want to have **CONSTANTS**. Static keyword doesn't make underlaying objects immutable and that's the problem. They can be changed at any time and lead to many weird issues and problems.

**Comment by ingmargoudt** [comment]
they are defined as final static, did we falsely use that as a definition for a constant?

**Comment by Zzooouhh** [comment]
@ingmargoudt What are the possible negative consequences of using == instead of .equals when comparing UUIDs? I've conducted dozens of tests with the card and it worked well with everything I threw at it. Could something break it this way?

**Comment by quercitron** [comment]
Having filters as they are now is like sitting on a bomb. One incorrect card implementation can break everything.
I agree that the best solution can be immutable filters. But we need to decide what is the most elegant way to add them to existing code.

**Comment by quercitron** [comment]
By the way, there was a similar problem with static mutable mana. On the whole server islands were producing black mana, or something like this. At the end code was modified to return new instance of mana every time, you can see details here https://github.com/magefree/mage/pull/1396.
But it is not a solution for filters problem.

**Comment by JayDi85** [comment]
@quercitron unit tests can catch that cards now. No more broken data by filters. But there are can be another pitfalls with other static/global data.

**Comment by quercitron** [comment]
Oh, I missed this commit. It's nice to have such check.

**Comment by LevelX2** [comment]
> @ingmargoudt What are the possible negative consequences of using == instead of .equals when comparing UUIDs?

equals checks if the value of the UUID is the same. and == checks if the UUID objects are the same.
Indeed there are situations where the UUID objects are copied/created newly for the same object and the "== " comparing fails although you don't intend that.
So always using equals is the way to go.

**Comment by ingmargoudt** [comment]
== does a check on memory reference, .equals checks on hash/content. Two objects can be the same without having the same memory reference

**Comment by JayDi85** [comment]
So, that's code with `==` UUID can be broken?

![shot_180113_142848](https://user-images.githubusercontent.com/8344157/34905251-39e05e28-f86e-11e7-9337-a4bb1b32a5db.png)

**Comment by LevelX2** [comment]
I completely agree, that the filter objects should be immutable. That would have the true advantage, that the filter objects wouldn't be copied during all kind of copy actions that take place during game play.
Also the StaticFilters we use now, if used e.g. in Targets of cards are copied all the time, if game states or abilities are copied (what happens a lot).
And if you need to have a filter modified, you always would create/get a new filter. That shouldn't raise problems for more than a handful of cards, that use filters in a special way.
Also often it felt wrong to me to create a new filter object while only needing another filter message. Maybe this is also something that could be taken care of, when creating a new filter object handling.

But no idea yet, if that's implementable without changing the use of filters in thousands of cards, creating a lot of work for the implementaion.

**Comment by Zzooouhh** [comment]
"equals checks if the value of the UUID is the same. and == checks if the UUID objects are the same."
Yes, but in cases like Raging River the code is literally seeking the same objects (players by UUID) - so wouldn't using '==' actually be slightly beneficial in mitigating the small risk of UUID collisions?

**Comment by JayDi85** [comment]
Created  #4407 for compare objects problem.

**Comment by LevelX2** [comment]
@Zzooouhh 
The problem is that you can have two different UUID objects for the same player (see JayDi85's #4407 for example).
So using == ends up saying it's a different player but it was the same player.
